optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/uni/UniAnchor.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.score.stream.drools.uni;
+
+import java.math.BigDecimal;
+import java.util.Arrays;
+import java.util.List;
+import java.util.UUID;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.ToIntFunction;
+import java.util.function.ToLongFunction;
+
+import org.drools.model.Declaration;
+import org.drools.model.Global;
+import org.drools.model.PatternDSL;
+import org.drools.model.RuleItemBuilder;
+import org.drools.model.consequences.ConsequenceBuilder;
+import org.kie.api.runtime.rule.RuleContext;
+import org.optaplanner.core.api.score.holder.AbstractScoreHolder;
+import org.optaplanner.core.impl.score.stream.bi.AbstractBiJoiner;
+import org.optaplanner.core.impl.score.stream.common.JoinerType;
+import org.optaplanner.core.impl.score.stream.drools.bi.BiAnchor;
+import org.optaplanner.core.impl.score.stream.drools.common.LogicalRuleMetadata;
+import org.optaplanner.core.impl.score.stream.drools.common.LogicalTuple;
+import org.optaplanner.core.impl.score.stream.drools.common.OriginalRuleMetadata;
+import org.optaplanner.core.impl.score.stream.drools.common.RuleMetadata;
+
+import static org.drools.model.DSL.on;
+
+public final class UniAnchor {
```

Comment:
Not sure about the name, `Anchor`.

The reason why this class exists is generics. `GroupingStream` operates on variables of `A`, but its children naturally need to operate on `LogicalTuple`. This was causing all kinds of trouble with inheritance, so I encapsulated all the generic mess into `Anchor` which - from the outside - is still as much type-safe as possible.

Now, the `Anchor`s do most of the work.

Reply:
-1 for using the name "Anchor" as it's confuses with chained variable's anchor.
Let's discuss what it is

Reply:
DroolsUniCondition?